### PR TITLE
B/time evicting queue

### PR DESCRIPTION
**Pull request by m-kharbat**

### CODE REVIEW

File: sql/src/test/java/io/crate/operation/reference/sys/TimeEvictingConcurrentLogQueueTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Test
+    public void testStopCleaningUp() throws Exception {
+        TimeEvictingConcurrentLogQueue queue = new TimeEvictingConcurrentLogQueue(TimeValue.timeValueSeconds(5));
+        queue.offer(new TestLog("expired", 2000L));
+        queue.offer(new TestLog("not expired", 8000L));
+        queue.offer(new TestLog("expired but added late", 3000L));
+        // The cleanup stops after the first not expired element
+        queue.removeExpired(9000);
```
Comment (mikethebeer) [inline sql/src/test/java/io/crate/operation/reference/sys/TimeEvictingConcurrentLogQueueTest.java]: `9000L` to be consistent

File: sql/src/main/java/io/crate/operation/reference/sys/Log.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.sys;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/Log.java]: I think we should move this into some other package. `operation.reference.sys` doesn't fit

File: sql/src/main/java/io/crate/operation/reference/sys/Log.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+public interface Log {
+
+    long started();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/Log.java]: `started` isn't required by the TimeEvictingConcurrentLogQueue.

File: sql/src/main/java/io/crate/operation/reference/sys/TimeEvictingConcurrentLogQueue.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package io.crate.operation.reference.sys;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class TimeEvictingConcurrentLogQueue extends ConcurrentLinkedQueue<Log> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/reference/sys/TimeEvictingConcurrentLogQueue.java]: Could use some javadocs.

File: sql/src/main/java/io/crate/operation/reference/sys/Log.java (hunk 2 above)
Comment (m-kharbat) [inline sql/src/main/java/io/crate/operation/reference/sys/Log.java]: where do you suggest that I move this interface ?

**Comment by seut** [comment]
@mfussenegger I don't like your suggestion of passing in a `Collection<Log>` because then one must ensure that the removal on the collection is safe on concurrent operations. It would make things complex at I see no reason for doing that.

Also we've already discussed how the removal/eviction should be triggered and doing it on write wouldn't be enough, so it would have to be done on read which increases the read latency enormous on large queues. We'd rather like to go for using an periodic scheduler currently.

@m-kharbat and me just agreed on moving the removeExpired method to the StatsTables class and let it operate on concrete ConcurrentLinkedQueue(s) so we can remove the interface and decouple the code.

**Comment by mfussenegger** [comment]
> @m-kharbat and me just agreed on moving the removeExpired method to the StatsTables class and let it operate on concrete ConcurrentLinkedQueue(s) so we can remove the interface and decouple the code.

sounds good to me.

>  don't like your suggestion of passing in a Collection<Log> because then one must ensure that the removal on the collection is safe on concurrent operations. It would make things complex and I see no reason for doing that.

It would be up to the caller to know if it's safe. The caller knows if the collection is used from multiple threads or not. If it needs to be synchronized, it can be done from outside.

Anyway, I think I phrased my suggestion a bit badly. My mentioning of  `Collection<Log>` was an example to show that it's unnecessary to subclass a Queue. Didn't intend to imply that `Collection` has to be used.

**Pull request closed without merging**